File: packages/browser-domain/src/social-policy-candidate-compiler.ts
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import { ParentTimestampSchema } from '@ocentra-parent/schema-domain/family-reference-primitives';
import { PolicyCompilerCapabilityState } from '@ocentra-parent/schema-domain/policy-compiler';
import {
  type SocialParentPolicyCompilerInput,
  SocialParentPolicyCompilerInputSchema,
  type SocialParentPolicyDecisionCandidate,
  SocialParentPolicyDecisionCandidateSchema,
} from '@ocentra-parent/schema-domain/social-policy-compiler';
import {
  SocialParentPolicyActionCandidateSchema,
  SocialParentPolicyConfidenceSchema,
  SocialParentPolicyDecisionCandidateIdSchema,
  SocialParentPolicyReasonCodesSchema,
} from '@ocentra-parent/schema-domain/social-policy-compiler-values';

const SocialParentPolicyCompileRequestSchema = withParser(
  Schema.Struct({
    input: SocialParentPolicyCompilerInputSchema,
    decisionCandidateId: SocialParentPolicyDecisionCandidateIdSchema,
    decidedAt: ParentTimestampSchema,
    expiresAt: Schema.Union(ParentTimestampSchema, Schema.Null),
    actionCandidate: SocialParentPolicyActionCandidateSchema,
    reasonCodes: SocialParentPolicyReasonCodesSchema,
    confidence: SocialParentPolicyConfidenceSchema,
    fallbackUsed: Schema.Boolean,
    parentApprovalRequired: Schema.Boolean,
  })
);

type SocialParentPolicyCompileRequest = Infer<typeof SocialParentPolicyCompileRequestSchema>;

export function compileSocialParentPolicyCandidate(
  request: SocialParentPolicyCompileRequest
): SocialParentPolicyDecisionCandidate {
  const parsed = SocialParentPolicyCompileRequestSchema.parse(request);
  const input = parsed.input;

  return SocialParentPolicyDecisionCandidateSchema.parse({
    schemaVersion: input.schemaVersion,
    decisionCandidateId: parsed.decisionCandidateId,
    compileRequestId: input.compileRequestId,
    decidedAt: parsed.decidedAt,
    expiresAt: parsed.expiresAt,
    policyVersionRef: input.policyVersionRef,
    targetKind: input.targetKind,
    sourceEvidenceRefs: input.sourceEvidenceRefs,
    signalSetRefs: input.signalSetRefs,
    parentRuleRefs: input.parentRuleRefs,
    scheduleContextRefs: input.scheduleContextRefs,
    timeBudgetContextRefs: input.timeBudgetContextRefs,
    scheduleState: input.scheduleState,
    timeBudgetState: input.timeBudgetState,
    actionCandidate: parsed.actionCandidate,
    reasonCodes: parsed.reasonCodes,
    confidence: parsed.confidence,
    compilerMode: input.compilerMode,
    fallbackUsed: parsed.fallbackUsed,
    parentApprovalRequired: parsed.parentApprovalRequired,
    compilerCapabilityState: compilerCapabilityStateForMode(input.compilerMode),
    finalPolicyDecisionClaimed: false,
    runtimeGateExecutedClaimed: false,
    uiRenderedClaimed: false,
    enforcementClaimed: false,
    nativeAppControlClaimed: false,
    platformConnectorClaimed: false,
    rawSignalPayloadStored: false,
    rawModelTextUsed: false,
  });
}

function compilerCapabilityStateForMode(mode: SocialParentPolicyCompilerInput['compilerMode']) {
  switch (mode) {
    case 'contract-only':
      return PolicyCompilerCapabilityState.Supported;
    case 'manual-required':
      return PolicyCompilerCapabilityState.ManualRequired;
    case 'unavailable':
      return PolicyCompilerCapabilityState.Unsupported;
  }
}

File: crates/browser-core/src/social_policy_candidate_compiler.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum SocialPolicyCompilerMode {
    ContractOnly,
    ManualRequired,
    Unavailable,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct SocialPolicyDecisionTemplate {
    pub compiler_capability_state: &'static str,
    pub final_policy_decision_claimed: bool,
    pub runtime_gate_executed_claimed: bool,
    pub ui_rendered_claimed: bool,
    pub enforcement_claimed: bool,
    pub native_app_control_claimed: bool,
    pub platform_connector_claimed: bool,
    pub raw_signal_payload_stored: bool,
    pub raw_model_text_used: bool,
}

const CONTRACT_ONLY_TEMPLATE: SocialPolicyDecisionTemplate = SocialPolicyDecisionTemplate {
    compiler_capability_state: "supported",
    final_policy_decision_claimed: false,
    runtime_gate_executed_claimed: false,
    ui_rendered_claimed: false,
    enforcement_claimed: false,
    native_app_control_claimed: false,
    platform_connector_claimed: false,
    raw_signal_payload_stored: false,
    raw_model_text_used: false,
};

const MANUAL_REQUIRED_TEMPLATE: SocialPolicyDecisionTemplate = SocialPolicyDecisionTemplate {
    compiler_capability_state: "manual-required",
    ..CONTRACT_ONLY_TEMPLATE
};

const UNAVAILABLE_TEMPLATE: SocialPolicyDecisionTemplate = SocialPolicyDecisionTemplate {
    compiler_capability_state: "unsupported",
    ..CONTRACT_ONLY_TEMPLATE
};

pub fn evaluate_social_policy_candidate(
    mode: SocialPolicyCompilerMode,
) -> SocialPolicyDecisionTemplate {
    match mode {
        SocialPolicyCompilerMode::ContractOnly => CONTRACT_ONLY_TEMPLATE,
        SocialPolicyCompilerMode::ManualRequired => MANUAL_REQUIRED_TEMPLATE,
        SocialPolicyCompilerMode::Unavailable => UNAVAILABLE_TEMPLATE,
    }
}

pub fn social_policy_candidate_compiler_typescript() -> String {
    include_str!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../browser-core-generated/social_policy_candidate_compiler.ts"
    ))
    .to_string()
}
